### DISCUSSION TITLE
Expand llama.cpp to llm.cpp

### DISCUSSION THREAD

**Discussion by howard0su** [ggml-org/llama.cpp#2102] [Ideas]
Make llama.cpp as a generic framework and SDK for LLM in C++.

## Goal

Expand the project scope to support additional LLM projects.

## API Definitions

Today's API is around llama but it can be generally divided into multiple categories:
1. Load/save model weights realted functions, which should be able to against any ggml binary file. largely the functions should not tight with llama. Model's tensor names can be different for the different models.
2. Tokenizer, we only support llama tokenizer today. The API needs to extend to support multi tokenizer. But the interface should kept same for any tokenizer. We need to support create_tokenizer, delete_toeknizer, id_to_token, token_to_id functions at least.
3. Samplers, we have many samplers. they are not tight with llama as well. we can make them a set of sampler functions. One small additional thing we should support predefined sampling pipelines.
4. Eval, the interface today is generic enough to support any model. The only change we need is extending init_parameter to support additional/different parameters for the different models. The different models will use different compute graph for batch and single token generation.
5. Finetuning/Training: Not implemented yet. We need to think more how to leverage Eval part like compute graph.
6. Offloading functions: We should be able to provide a set of classes to offload the tensor to GPU and graph computing logic will leverage the related backend support to do the compute. Offload functionality can be implemented as an interface.

## Workitem

1. First come first, rename to LLM.cpp :)
2. Split the interface into llama specific and LLM general functions.
3. Refactor and decouple the current code especially context, model as the connection point.
4. Adopt other LLM models.

**Comment by Green-Sky** [comment]
### api:
1. ties into [gguf](https://github.com/ggerganov/ggml/pull/302)
3. they are pretty standalone already
5. @xaedes has made good progress on that
6. the move to graph compute kind of already does this (also ties into gguf)

### workitem:
pretty sure 1. comes afer 4. :laughing:

**Comment by ianscrivener** [comment]
IMO changing the "llama.cpp" would be bad idea as "llama.cpp" has a LOT of name recognition, prior publicity and traction. (I believe Mark Zuckerberg name dropped 'llama.cpp')

I totally get that your goal @howard0su is to move the code base beyond just handling 'llama family' LLM models... but IMO we should stick with the current (famous) name.

**Comment by lin72h** [comment]
like the name **llm.cpp** a lot. minimalistic , right to the point, very **gg**
